scipy/sparse/linalg/_isolve/tests/test_iterative.py

```diff
@@ -538,6 +538,23 @@ def test_x0_solves_problem_exactly(solver):
     assert info == 0
 
 
+@pytest.mark.parametrize("solver", _SOLVERS)
+@pytest.mark.parametrize("n", [1, 3, 11])
+def test_shape_consistency(solver, n):
+    # See gh-21164
+    # 1D array
+    vals = np.arange(1, n + 1, dtype=float)
+    mat = np.diag(vals)
+    # 2D array
+    vals2 = vals[:, np.newaxis]
+    for rhs in [vals, 0. * vals, vals2, 0. * vals2]:
+        for x0 in [vals, vals2]:
+            sol, _ = solver(mat, rhs)
+            assert_array_equal(np.shape(sol), np.shape(rhs))
+            sol, _ = solver(mat, rhs, x0)
```

Comment:
This runs 30 test cases, which is probably too heavy for a simple bug. Please could we thin this out a bit, maybe like the following
```suggestion
@pytest.mark.parametrize("solver", _SOLVERS)
def test_shape_consistency(solver):
    # See gh-21164
    # 1D array
    vals = np.arange(1, 5)
    mat = np.diag(vals)
    # 2D array
    vals2 = vals[:, np.newaxis]
    for rhs in [vals, vals2]:
        sol, _ = solver(mat, rhs)
        assert_array_equal(np.shape(sol), np.shape(rhs))
```

Reply:
I have not looked at all codes of all solvers.

1 and 3 may be different edge cases, 10 should I guess be generic. You version only checks 5. In the current code, the fix is done in postprocessing common for all cases, so it makes no difference what or how many cases to check. My check is future proof.